Refactor editing dashboard's cards

qnkhuat opened the pull request:
Milestone 0 of https://github.com/metabase/metabase/issues/29502

### Motivation
Currently, the APIs we used to create and delete dashcards are not bulked, which means if you edit a dashboard and add 5 cards to it, FE will calls `POST /dashboard/cards` 5 times.

That's not cool and we want to do bulk operations instead. So to do that we'll Remove 
- `DELETE  /api/dashboard/:id/cards` 
- `POST /api/dashboard/:id/cards` 

Update `PUT /api/dashboard/:id/cards` to handle bulks create/update/delete at the same time.
it takes a payload with keys `cards`
- to create cards, include a card with a negative id
- to delete cards, don't include that card in the payload
- to update cards, update card information in the cards payload

FE PR (already merged onto this branch): https://github.com/metabase/metabase/pull/29762

### Demo

https://user-images.githubusercontent.com/37751258/231255437-1ab4e592-c3c3-44b6-baf5-a36579139549.mov

Adding and removing cards and confirming revision history is correct.



<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/29641)
<!-- Reviewable:end -->

qnkhuat committed:
BE: MVP to bulk create/delete APIs

qnkhuat committed:
tests fixes

qnkhuat committed:
test fixes

qnkhuat committed:
use bulk cards insert everywhere

qnkhuat committed:
bulk everything -- add dashcards

qnkhuat committed:
bulk everything for real now(previously only bulk for post/update cards)

qnkhuat committed:
test fixes

qnkhuat committed:
Overload POST/DELETE into PUT /dashboard/:id/cards

qnkhuat committed:
Merge branch 'master' into dashboard-bulk-create-delete

qnkhuat committed:
tests fixes

qnkhuat committed:
rearrange tests

qnkhuat committed:
remove inline defs and fix kondo

qnkhuat committed:
fix can't prune cards

qnkhuat committed:
use malli for schema

qnkhuat committed:
generalize the classify changes function

qnkhuat committed:
Merge branch 'master' into dashboard-bulk-create-delete

qnkhuat committed:
optimize DB calls

qnkhuat committed:
makes sure we dissoc id when create cards and makes side-affects function end with exclaimation

qnkhuat committed:
makes sure we dissoc id when create cards and makes side-affects function end with exclaimation

EmmadUsmani committed:
refactor `saveDashboardAndCards` thunk to use new `PUT` endpoint for dashcards (#29762)

qnkhuat commented on src/metabase/api/dashboard.clj:
```diff
@@ -539,24 +513,67 @@
                                          (assoc mapping :card-id (get dashcard-id->card-id dashcard-id)))]
     (check-parameter-mapping-permissions new-mappings)))
 
+(mu/defn ^:private classify-changes :- [:map
```
I made this generic, so that I can later on use this for tabs as well.

qnkhuat commented on test/metabase/api/dashboard_test.clj:
this file has lots of changes because I reordered some tests to put it into the correct sections.

I commented on which tests are new.

qnkhuat commented on test/metabase/api/dashboard_test.clj:
```diff
@@ -1342,57 +1240,220 @@
                                                              {:cards [new-dashcard-info]}))}))))))
 
-(deftest update-cards-parameter-mapping-permissions-test
+(deftest e2e-update-cards-test
```
new test

qnkhuat commented on test/metabase/api/dashboard_test.clj:
```diff
@@ -1474,15 +1556,70 @@
 
+;;; -------------------------------------- Delete dashcards only tests ---------------------------------------
+
+(deftest delete-cards-test
```
new test

qnkhuat commented on test/metabase/api/dashboard_test.clj:
```diff
@@ -1474,15 +1556,70 @@
+          (is (= 0
+                 (count (t2/select-pks-set DashboardCard, :dashboard_id dashboard-id)))))))))
+
+(deftest classify-changes-test
```
new test

metamben commented on src/metabase/api/dashboard.clj:
```diff
@@ -539,24 +513,67 @@
+  (for [{:keys [card_id]} cards
+        :when  card_id]
+    (api/check-not-archived (api/read-check Card card_id)))
+  (check-parameter-mapping-permissions (flatten (for [{:keys [card_id parameter_mappings]} cards]
+                                                  (for [mapping parameter_mappings]
+                                                    (assoc mapping :card-id card_id)))))
```
```suggestion
  (check-parameter-mapping-permissions (for [{:keys [card_id parameter_mappings]} cards
                                             mapping parameter_mappings]
                                         (assoc mapping :card-id card_id)))
```

qnkhuat replied:
awesome, forgot that we could have nested for destructring

qnkhuat replied:
0d7da010b2

metamben commented on src/metabase/api/dashboard.clj:
```diff
@@ -539,24 +513,67 @@
+                                                    (assoc mapping :card-id card_id)))))
+  (u/prog1 (api/check-500 (dashboard/add-dashcards! dashboard (map #(assoc % :creator_id @api/*current-user*) cards)))
+    (events/publish-event! :dashboard-add-cards {:id (:id dashboard) :actor_id api/*current-user-id* :dashcards <>})
+    (doseq [{:keys [card_id]} cards]
```
These `card_id`s are all dummies, aren't they? Should we iterate over `<>`? And if I'm not mistaken, then some tests are missing or wrong.

qnkhuat replied:
yep it should, will change,

metamben commented on src/metabase/api/dashboard.clj:
```diff
@@ -539,24 +513,67 @@
+  ;; transform the card data to the format of the DashboardCard model
+  ;; so update-dashcards! can compare them with existing cards
+  (dashboard/update-dashcards! dashboard (map dashboard-card/from-parsed-json cards))
+  (events/publish-event! :dashboard-reposition-cards {:id (:id dashboard) :actor_id api/*current-user-id* :dashcards cards}))
```
Is the position the only thing that can change?

qnkhuat replied:
this is from before so I'm not sure, but I guess that what we wanted to check.

metamben commented on src/metabase/models/dashboard_card.clj:
```diff
@@ -135,24 +137,26 @@
+  *  All cards will be updated with a `position` according to their place in the collection of `card-ids`"
+  {:arglists '([dashcard-id->card-ids])}
+  [dashcard-id->card-ids]
+  (when-not (empty? dashcard-id->card-ids)
```
```suggestion
  (when (seq dashcard-id->card-ids)
```

qnkhuat replied:
0d7da010b2

metamben commented on src/metabase/models/dashboard_card.clj:
```diff
@@ -135,24 +137,26 @@
+    ;; first off, just delete all series on the dashboard card (we add them again below)
+    (t2/delete! DashboardCardSeries :dashboardcard_id [:in (keys dashcard-id->card-ids)])
+    ;; now just insert all of the series that were given to us
+    (when-let [card-series (->> (for [[dashcard-id card-ids] dashcard-id->card-ids]
+                                  (map-indexed (fn [i card-id]
+                                                 {:dashboardcard_id dashcard-id, :card_id card-id, :position i})
+                                               card-ids))
+                                flatten
+                                (remove nil?))]
+      (t2/insert! DashboardCardSeries card-series))))
```
1. `remove` cannot return a falsy value, so `when-let` is just `let`.
2. I'm not sure how `nil`s can show up in the list at all, we seem to be putting maps there.
3. I think we can get rid of `flatten` by continuing the list comprehension.
```suggestion
    (when-let [card-series (seq (for [[dashcard-id card-ids] dashcard-id->card-ids
                                      [i card-id] (map-indexed vector card-ids)]
                                  {:dashboardcard_id dashcard-id, :card_id card-id, :position i}))]
      (t2/insert! DashboardCardSeries card-series))))
```

qnkhuat replied:
This is much better and your assessment is correct. Thanks

qnkhuat replied:
0d7da010b2

metamben commented on src/metabase/models/dashboard_card.clj:
```diff
@@ -201,48 +205,45 @@
+  DashboardCardSeries. Returns the newly created DashboardCard or throws an Exception."
+  [dashboard-cards :- [:sequential NewDashboardCard]]
+  (t2/with-transaction [_conn]
+    (let [dashboard-card-ids (t2/insert-returning-pks!
```
We usually check if the input is empty, but not in this case. Is that intentional?

qnkhuat replied:
not intentional, I'll add a seq check.

qnkhuat replied:
0d7da010b2

metamben commented on test/metabase/api/dashboard_test.clj:
```diff
@@ -1342,57 +1240,220 @@
+                   updated-card-2
+                   new-card]
+                  resp))
+          ;; dashboard 3 is deleted
```
```suggestion
          ;; dashcard 3 is deleted
```

metamben reviewed (commented):
I wonder if the subtle changes in the tests (size of cards, creating cards in one step instead of two (create + update)) can cause or hide problems.

BE looks good overall, but I would like to see your answers to some of my questions before approving.

qnkhuat commented on src/metabase/api/dashboard.clj:
```diff
@@ -539,24 +513,67 @@
+                                        [:to-create [:sequential [:map [:id neg-int?]]]]
+                                        [:to-update [:sequential [:map [:id ms/PositiveInt]]]]
+                                        [:to-delete [:sequential [:map [:id ms/PositiveInt]]]]]
+  [current-changes :- [:sequential [:map [:id ms/PositiveInt]]]
+   new-changes     :- [:sequential [:map [:id int?]]]]
```
I intentionally made this generic because it's gonna be used for "tabs"(a new entity) later on.

qnkhuat commented on src/metabase/api/dashboard.clj:
```diff
@@ -539,24 +513,67 @@
+  [current-changes :- [:sequential [:map [:id ms/PositiveInt]]]
+   new-changes     :- [:sequential [:map [:id int?]]]]
+  (let [current-dashcard-ids          (set (map :id current-changes))
+        update-or-delete-dashcard-ids (->> new-changes
```
could you explain why? because the list contains both ids for `update` and `delete`

qnkhuat commented on src/metabase/api/dashboard.clj:
```diff
@@ -539,24 +513,67 @@
     (check-parameter-mapping-permissions new-mappings)))
 
+(mu/defn ^:private classify-changes :- [:map
+                                        [:to-create [:sequential [:map [:id neg-int?]]]]
+                                        [:to-update [:sequential [:map [:id ms/PositiveInt]]]]
```
you mean between `to-create` and others? 

because negative ids in `new-changes` is considered as `to-create` entity.

I added a docstring to this helper, hope it makes thing clearer.

qnkhuat replied:
docstring added here: 0d7da010b2

qnkhuat commented on src/metabase/api/dashboard.clj:
```diff
@@ -539,24 +513,67 @@
+
+(defn- create-cards!
+  [dashboard cards]
+  (for [{:keys [card_id]} cards
+        :when  card_id]
+    (api/check-not-archived (api/read-check Card card_id)))
```
you're right that it was not tested. I'll add a test.

qnkhuat replied:
eb0bf80c48 fixed and add test here.

qnkhuat committed:
resolve Tamas's comments

qnkhuat committed:
add archive check test

qnkhuat commented:
@metamben 
> I wonder if the subtle changes in the tests (size of cards, creating cards in one step instead of two (create + update)) can cause or hide problems.

I'm not sure what is your concern, could you elaborate?

qnkhuat committed:
unwrap message from failed test

qnkhuat committed:
test fixes 🤨

EmmadUsmani committed:
fix e2e tests

EmmadUsmani commented on e2e/support/commands/api/composite/createNativeQuestionAndDashboard.js:
```diff
@@ -5,14 +5,22 @@ Cypress.Commands.add(
+            cy.request("PUT", `/api/dashboard/${dashboardId}/cards`, {
+              cards: [
+                {
+                  id: -1,
```
In [`dashboard/actions/data-fetching.js`](https://github.com/metabase/metabase/pull/29641/files#diff-636074b28010ac8c8f8597b27dc79a72530c430dbfd344340c399de22a1f6c80R82), the id for new dashcards used to be a random float between 0 and 1. In this PR we changed it to be negative numbers.

For new dashcards that the user has added while in edit mode, we need to have a unique id for each one so the client can differentiate them. This is why we can't use a single value like `null`, or a static descriptor like `"new"`. We also don't want to change the type of this value from a `number` to something else like a `string`, because doing so would break any typescript code that expects the dashcard id to be a `number`.

I agree it's not intuitive, but I don't think this is an issue. We don't ever manually construct and send the api request to add new dashcards, it's always done by dispatching a [`ADD_CARD_TO_DASH` redux action](https://github.com/metabase/metabase/blob/92a094159958a81ff0076058c826dbeb8cc3d011/frontend/src/metabase/dashboard/actions/cards.js#L55), which handles setting the id for you. 

In e2e I've refactored everything to use one of the helper methods that sends the request. I supposed we could remove it from the e2e code, but I'd rather keep it the same as how it's being done in the client, especially since the API expects that value to be populated.

EmmadUsmani commented on e2e/test/scenarios/dashboard-filters/reproductions/12720-no-data-permissions-connected-filter.cy.spec.js:
```diff
@@ -42,53 +47,39 @@ describe("issue 12720", () => {
+                },
+              ],
+            },
+          ],
```
The new test is actually creating one new card and updating the existing one. In the `PUT` request that's being sent, the dashcard with `card_id` `SQL_ID` is being created, since it was not on the dashboard already, while the card with `card_id` `1` is being updated, since it already was on the dashboard. 

This is probably unclear due to the name of the method, perhaps it should be renamed to `addOrUpdateDashCards`?

EmmadUsmani commented on e2e/test/scenarios/dashboard-filters/reproductions/25322-loading-list-values.cy.spec.js:
```diff
@@ -44,35 +49,19 @@ const createDashboard = () => {
-            });
-          });
+            },
+          }).then(() => ({ dashboard_id }));
```
I don't believe this is what the test case was trying to test. The original [issue](https://github.com/metabase/metabase/issues/25322) (and assertion in this test) was that there should be a loading spinner while filter values are being loaded. This should happen regardless of if the filter was added to the dashcard during its initial creation or later.

It was a common pattern throughout our e2e tests to manually send a `POST` request to create the dashcard, then a `PUT` request to link a filter to it. Why was this the case? I'm not entirely sure, but I would assume it had to do with some limitation of the old `POST` endpoint. I read through the test cases and none of them seemed to be explicitly testing first creating then later adding the filter to the dashcard.

For this reason, I think it's safe to change it to doing it in the single `PUT` request. All the test cases still pass, the code is cleaner, and performance should be better since we are not making an additional roundtrip to the server.

EmmadUsmani commented on e2e/test/scenarios/dashboard/chained-filters.cy.spec.js:
```diff
@@ -157,12 +158,9 @@ describe("scenarios > dashboard > chained filter", () => {
-          row: 0,
-          col: 0,
-          size_x: 8,
-          size_y: 6,
```
Yup, so in most places as I changed from manually sending the request to using the helper I tried to remove the explicit `size_x` and `size_y` definition, since usually it wasn't necessary (the tests worked fine with an 8x8 card). There were a few cases where it did need a specific size so I left those unchanged.

EmmadUsmani commented:
@qnkhuat I think @metamben is referring to the e2e tests I updated, where some of these things were changed as I refactored.

@metamben I appreciate the very thorough review! I run every test that I updated on my machine and confirmed that the behavior was the same as before and that the test passed. That in combination with running on CI should give us confidence that things aren't broken.

metamben reviewed (changes requested):
> Reviewed 2 of 8 files at r3, 4 of 7 files at r4, 36 of 37 files at r5, 1 of 3 files at r6, 1 of 2 files at r7, 1 of 1 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29641)* status: all files reviewed, 8 unresolved discussions (waiting on @EmmadUsmani and @qnkhuat)

___
*[`e2e/test/scenarios/dashboard-filters/reproductions/12720-no-data-permissions-connected-filter.cy.spec.js` line 82 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160926234:-NSgEBHV76IMMsoapyu-:b-zds1zv) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/e2e/test/scenarios/dashboard-filters/reproductions/12720-no-data-permissions-connected-filter.cy.spec.js#L82)):*
<details><summary><i>Previously, EmmadUsmani (Emmad Usmani) wrote…</i></summary><blockquote>

The new test is actually creating one new card and updating the existing one. In the `PUT` request that's being sent, the dashcard with `card_id` `SQL_ID` is being created, since it was not on the dashboard already, while the card with `card_id` `1` is being updated, since it already was on the dashboard. 

This is probably unclear due to the name of the method, perhaps it should be renamed to `addOrUpdateDashCards`?
</blockquote></details>

I was confused by the removed `id` of the first card.
___
*[`e2e/test/scenarios/dashboard-filters/reproductions/25322-loading-list-values.cy.spec.js` line 64 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160928149:-NSgF2Ah2m1f9PA33Ifa:buqqllt) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/e2e/test/scenarios/dashboard-filters/reproductions/25322-loading-list-values.cy.spec.js#L64)):*
>For this reason, I think it's safe to change it to doing it in the single PUT request. All the test cases still pass, the code is cleaner, and performance should be better since we are not making an additional roundtrip to the server.

This can well be the case even when we stop checking something important, but NWM, I trust your judgement.
___
*[`src/metabase/api/dashboard.clj` line 518 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160947579:-NSgFsoQ6HYdxq1ZL1nK:bhe6r6g) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L518)):*
<details><summary><i>Previously, qnkhuat (Ngoc Khuat) wrote…</i></summary><blockquote>

docstring added here: 0d7da010b2
</blockquote></details>

I meant `neg-int?` vs `ms/PositiveInt`.
___
*[`src/metabase/api/dashboard.clj` line 521 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160950697:-NSgG3l5DXjnRc1EFsbV:b8qq3do) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L521)):*
<details><summary><i>Previously, qnkhuat (Ngoc Khuat) wrote…</i></summary><blockquote>

I intentionally made this generic because it's gonna be used for "tabs"(a new entity) later on.
</blockquote></details>

I see. Still, `change` is not something I recognize as covering both "cards" and "tabs". I have a problem contrasting "current-change" with "new-change".

(Also, "card" is used later on, so I had no idea what you had in mind.)
___
*[`src/metabase/api/dashboard.clj` line 523 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160951380:-NSgGjMgDFt0y8DZ5gXz:b-l06yha) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L523)):*
<details><summary><i>Previously, qnkhuat (Ngoc Khuat) wrote…</i></summary><blockquote>

could you explain why? because the list contains both ids for `update` and `delete`
</blockquote></details>

how we construct `to-delete` (see the lines below) is confusing. if we `remove` (could be used instead of `filter` (and `comp` would help readability too, I guess)) the cards that are to be updated or deleted, then we why do we get the ones that should be deleted?



_Code quote:_
```Clojure
        to-delete (filter #(not (update-or-delete-dashcard-ids (:id %)))
                          current-changes)
```
___
*[`src/metabase/api/dashboard.clj` line 557 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160966979:-NSgIOHZ4siez9r0ZXCw:b9dewl5) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L557)):*
<details><summary><i>Previously, qnkhuat (Ngoc Khuat) wrote…</i></summary><blockquote>

this is from before so I'm not sure, but I guess that what we wanted to check.
</blockquote></details>

I mean, is it OK to send a reposition event when in fact the position hasn't changed?


<!-- Sent from Reviewable.io -->

metamben reviewed (commented):
@qnkhuat, @EmmadUsmani is right, I meant the e2e tests.

@EmmadUsmani just to clarify, I'm not a FE developer (and have only rudimentary JS skills), so you should get a stamp from a FE developer too. As for having run the tests, my concern was that we might have _relaxed_ some tests, so the fact that they all pass doesn't mean nothing is broken and that we haven't lost some important checks. But anyway, I trust your decision, just wanted to make sure these changes are intentional.

> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29641#-:-NSgJxsV0YZXjgzFLEv4:b-j2zcl4)* status: all files reviewed, 8 unresolved discussions (waiting on @EmmadUsmani and @qnkhuat)



<!-- Sent from Reviewable.io -->

EmmadUsmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29641)* status: all files reviewed, 8 unresolved discussions (waiting on @qnkhuat)

___
*[`e2e/test/scenarios/dashboard-filters/reproductions/12720-no-data-permissions-connected-filter.cy.spec.js` line 82 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160926234:-NSgO7cWCSD1HdlWuZpW:bkyj5sf) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/e2e/test/scenarios/dashboard-filters/reproductions/12720-no-data-permissions-connected-filter.cy.spec.js#L82)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

I was confused by the removed `id` of the first card.
</blockquote></details>

Ah that's a good catch actually, I think the `id` should still be there. Let me fix that.
___
*[`e2e/test/scenarios/dashboard-filters/reproductions/25322-loading-list-values.cy.spec.js` line 64 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160928149:-NSgOIk31hFReDq9ZWjj:bdwppdt) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/e2e/test/scenarios/dashboard-filters/reproductions/25322-loading-list-values.cy.spec.js#L64)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

>For this reason, I think it's safe to change it to doing it in the single PUT request. All the test cases still pass, the code is cleaner, and performance should be better since we are not making an additional roundtrip to the server.

This can well be the case even when we stop checking something important, but NWM, I trust your judgement.
</blockquote></details>

Right that makes sense to me. I think my main point (apologies if it was unclear) is that the separate `POST` and `PUT` requests was never what we wanted to check to begin with.


<!-- Sent from Reviewable.io -->

EmmadUsmani committed:
fix e2e test

alxnddr commented on e2e/support/helpers/e2e-dashboard-helpers.js:
```diff
@@ -10,6 +10,43 @@ export function getDashboardCard(index = 0) {
+    }));
+}
+
+export function addCardsToDashboard({ dashboard_id, cards }) {
```
nit: wdyt about naming it like `updateDashboardCards` and adding a jsdoc describing it is suitable for cards CRUD since `addCardsToDashboard`  makes it not so obvious it can be used for updating/removing cards

EmmadUsmani replied:
Yup that sounds better to me, I can rename it

EmmadUsmani replied:
It's been renamed on the latest commit

alxnddr commented on frontend/src/metabase/dashboard/actions/save.js:
```diff
@@ -150,15 +116,14 @@ export const saveDashboardAndCards = createThunkAction(
-      }
+        })),
+      });
+      dispatch(
```
Would it be simpler to dispatch a UPDATE_DASHCARDS with `updatedDashCards` which will update entire dashcards in the state instead of trying to update ids only??

EmmadUsmani replied:
Yeah I think that would be simpler, it will also help later when removing tabs as it will make it easy to update the dashcards in the removed tab (either delete them or update their tab id). I will make this change.

EmmadUsmani replied:
After looking into it I don't think we can do this. Here we are updating the `dashcardData` slice, not `dashcards`, and which are two different types. Since we don't get the `dashcardData` objects in the api response, the best we can do is map between the old ids and the new ids like are doing here.

<img width="552" alt="Screenshot 2023-04-12 at 1 07 33 PM" src="https://user-images.githubusercontent.com/37751258/231573574-42b3ec0d-e1d8-4c86-a910-a5c5edeea59b.png">

alxnddr replied:
@EmmadUsmani makes sense, thanks for looking at it anyway

alxnddr reviewed (approved):
Looks great! Please have a look at my comment before merging

alxnddr reviewed (changes requested):
The revision history is not computed correctly.
If you have a dashboard with 3 cards, then remove one and add another one it will show you just rearranged cards

qnkhuat committed:
refactor the classify-changes to make it more readable

qnkhuat reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29641)* status: 46 of 47 files reviewed, 10 unresolved discussions (waiting on @EmmadUsmani and @metamben)

___
*[`src/metabase/api/dashboard.clj` line 518 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160947579:-NSj75MI-Z38k5aCY3IB:b-fblq4x) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L518)):*
> I meant neg-int? vs ms/PositiveInt.
as I described in the docstring, FE will send over a list of cards, to classify between create versus update-or-delete, we have a convention that card with negative ids are for creation purpose.
___
*[`src/metabase/api/dashboard.clj` line 521 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160950697:-NSjJjk4BtQ7QBkiu6fd:bmnv55m) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L521)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

I see. Still, `change` is not something I recognize as covering both "cards" and "tabs". I have a problem contrasting "current-change" with "new-change".

(Also, "card" is used later on, so I had no idea what you had in mind.)
</blockquote></details>

I agree but I'm not sure i have a better name, suggestions are welcomed :pray:
___
*[`src/metabase/api/dashboard.clj` line 523 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160951380:-NSjJebh31HPjs0pfBCe:bbgen2q) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L523)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

how we construct `to-delete` (see the lines below) is confusing. if we `remove` (could be used instead of `filter` (and `comp` would help readability too, I guess)) the cards that are to be updated or deleted, then we why do we get the ones that should be deleted?
</blockquote></details>

yea that does make sense, I'll refactor this, can you take another look? 86050e702f 
___
*[`src/metabase/api/dashboard.clj` line 557 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160966979:-NSjIz5393HI7ho50WQx:b3o79gw) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L557)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

I mean, is it OK to send a reposition event when in fact the position hasn't changed?
</blockquote></details>

I see, I'm not too familiar with the revision atm so I'll not change it now. 


<!-- Sent from Reviewable.io -->

qnkhuat committed:
fixes failed to compile

EmmadUsmani committed:
fix revision history on FE

metamben reviewed (changes requested):
> Reviewed 1 of 1 files at r10, 2 of 2 files at r11, 2 of 2 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29641)* status: all files reviewed, 9 unresolved discussions (waiting on @alxnddr and @qnkhuat)

___
*[`src/metabase/api/dashboard.clj` line 518 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160947579:-NSm17E58jwhGeBZr1Ht:b5d7yn3) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L518)):*
<details><summary><i>Previously, qnkhuat (Ngoc Khuat) wrote…</i></summary><blockquote>

> I meant neg-int? vs ms/PositiveInt.
as I described in the docstring, FE will send over a list of cards, to classify between create versus update-or-delete, we have a convention that card with negative ids are for creation purpose.
</blockquote></details>

I meant the asymmetry `neg-int?` on one hand `ms/PositiveInt` on the other. Why not `pos-int?`. Anyway, this is an absolute nit.
___
*[`src/metabase/api/dashboard.clj` line 521 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160950697:-NSm1V0o6Tq53rFdWKhP:b-31odrw) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L521)):*
<details><summary><i>Previously, qnkhuat (Ngoc Khuat) wrote…</i></summary><blockquote>

I agree but I'm not sure i have a better name, suggestions are welcomed :pray:
</blockquote></details>

`items`?
___
*[`src/metabase/api/dashboard.clj` line 523 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160951380:-NSm1r7V9alPJ_fShk04:b25f3pv) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L523)):*
<details><summary><i>Previously, qnkhuat (Ngoc Khuat) wrote…</i></summary><blockquote>

yea that does make sense, I'll refactor this, can you take another look? 86050e702f 
</blockquote></details>

much cleaner, thanks! (still, I suggest using `(comp current-change-ids :id)` and `remove` instead of `filter #(not ...)`.)
___
*[`src/metabase/api/dashboard.clj` line 561 at r12](https://reviewable.io/reviews/metabase/metabase/29641#-NSm2bIc-gm5gKDauOrc:-NSm2bId9oeaAzRPz7F3:bgog7z2) ([raw file](https://github.com/metabase/metabase/blob/9259ab493d166952982068e622d6c9a2701bcf54/src/metabase/api/dashboard.clj#L561)):*
> ```Clojure
>   ;; so update-dashcards! can compare them with existing cards
>   (dashboard/update-dashcards! dashboard (map dashboard-card/from-parsed-json cards))
>   ;; TODO this is ambiguous, we don't know for sure here that the cards are repositioned
> ```

"ambiguous" -> "potentially misleading"


<!-- Sent from Reviewable.io -->

EmmadUsmani committed:
update e2e helper names

alxnddr reviewed (approved):
FE looks good 🏅

qnkhuat reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29641)* status: 22 of 50 files reviewed, 9 unresolved discussions (waiting on @alxnddr, @EmmadUsmani, and @metamben)

___
*[`src/metabase/api/dashboard.clj` line 518 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160947579:-NSsEGUk9DB0e6PPoGFM:b-v7n6rt) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L518)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

I meant the asymmetry `neg-int?` on one hand `ms/PositiveInt` on the other. Why not `pos-int?`. Anyway, this is an absolute nit.
</blockquote></details>

Using schema defined from `ms` is usually better because it's wrapped with a nice error message. I created a new `ms/NegativeInt` and use it here d894d8f2cf
___
*[`src/metabase/api/dashboard.clj` line 521 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160950697:-NSsE9KH0uBefpTaiMjc:b-gftzjn) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L521)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

`items`?
</blockquote></details>

how about `classify-items-change`?
___
*[`src/metabase/api/dashboard.clj` line 523 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160951380:-NSsES-X9TJpPXIFlWNi:b-h6cj7o) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L523)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

much cleaner, thanks! (still, I suggest using `(comp current-change-ids :id)` and `remove` instead of `filter #(not ...)`.)
</blockquote></details>

updated d894d8f2cf
___
*[`src/metabase/api/dashboard.clj` line 561 at r12](https://reviewable.io/reviews/metabase/metabase/29641#-NSm2bIc-gm5gKDauOrc:-NSsEFGE6lPcSH9UWdXq:b-w18g8t) ([raw file](https://github.com/metabase/metabase/blob/9259ab493d166952982068e622d6c9a2701bcf54/src/metabase/api/dashboard.clj#L561)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

"ambiguous" -> "potentially misleading"
</blockquote></details>

updated in d894d8f2cf


<!-- Sent from Reviewable.io -->

qnkhuat reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29641)* status: 22 of 50 files reviewed, 9 unresolved discussions (waiting on @alxnddr, @EmmadUsmani, and @metamben)

___
*[`src/metabase/api/dashboard.clj` line 521 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160950697:-NSsEcJF2jJ8l1safzcW:b-lh03j1) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L521)):*
<details><summary><i>Previously, qnkhuat (Ngoc Khuat) wrote…</i></summary><blockquote>

how about `classify-items-change`?
</blockquote></details>

I like the word `change` because it tells that the output is used to `change` or `modify` something.


<!-- Sent from Reviewable.io -->

calherries reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29641)* status: 22 of 50 files reviewed, 16 unresolved discussions (waiting on @alxnddr, @EmmadUsmani, @metamben, and @qnkhuat)

___
*[`src/metabase/api/dashboard.clj` line 521 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160950697:-NSu65jj7kKS-NbO3oD0:be4ve7t) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L521)):*
<details><summary><i>Previously, qnkhuat (Ngoc Khuat) wrote…</i></summary><blockquote>

I like the word `change` because it tells that the output is used to `change` or `modify` something.
</blockquote></details>

How about `diff-by-id`, since it's basically doing a `clojure.data.diff` on the ids
___
*[`src/metabase/api/dashboard.clj` line 535 at r14](https://reviewable.io/reviews/metabase/metabase/29641#-NSu3Vft5QWa12tvcTZe:-NSu3Vft5QWa12tvcTZf:bttxow0) ([raw file](https://github.com/metabase/metabase/blob/d894d8f2cf98911921dde0ca94962113a77e1b34/src/metabase/api/dashboard.clj#L535)):*
> ```Clojure
>         to-update          (filter #(current-change-ids (:id %)) new-changes)
>         ;; to delete changes in current but not new changes
>         to-delete          (remove (comp new-change-ids :id) current-changes)]
> ```

Perhaps this is simpler to understand if you used `clojure.data.diff` on the `:id`.

e.g.
```
(let ...
[create-ids delete-ids update-ids] (diff new-change-ids current-change-ids)
...)
```



_Code quote:_
```Clojure
        to-create          (filter #(neg? (:id %)) new-changes)
        ;; to-update changes are new changes with id in the current changes
        to-update          (filter #(current-change-ids (:id %)) new-changes)
        ;; to delete changes in current but not new changes
        to-delete          (remove (comp new-change-ids :id) current-changes)]
```
___
*[`src/metabase/api/dashboard.clj` line 540 at r14](https://reviewable.io/reviews/metabase/metabase/29641#-NSu6iXA0uqgltt-S-YQ:-NSu6iXA0uqgltt-S-YR:b-getgit) ([raw file](https://github.com/metabase/metabase/blob/d894d8f2cf98911921dde0ca94962113a77e1b34/src/metabase/api/dashboard.clj#L540)):*
> ```Clojure
>      :to-create to-create}))
> 
> (defn- create-cards!
> ```

How about renaming this to `create-dashcards`? I think it's passable to use `cards` instead of `dashcards` in a context when it's not ambiguous. But top-level functions lack context by themselves. If you have functions calling functions calling functions where everything is named `card` and not `dashcard`, I'd argue it slows down reading because you have to check whether a thing called `card` is really a `card` or a `dashcard`.
___
*[`src/metabase/api/dashboard.clj` line 541 at r14](https://reviewable.io/reviews/metabase/metabase/29641#-NSu8OSg1nJQqRjdawj9:-NSu8OSg1nJQqRjdawjA:b-1pfpac) ([raw file](https://github.com/metabase/metabase/blob/d894d8f2cf98911921dde0ca94962113a77e1b34/src/metabase/api/dashboard.clj#L541)):*
> ```Clojure
> 
> (defn- create-cards!
>   [dashboard cards]
> ```

Also would be nice to use `dashcards` instead of `cards`



_Suggestion:_
```Clojure
  [dashboard dashcards]
```
___
*[`src/metabase/api/dashboard.clj` line 558 at r14](https://reviewable.io/reviews/metabase/metabase/29641#-NSu8J7v43V-WdwlHZk9:-NSu8J7v43V-WdwlHZkA:b2u025a) ([raw file](https://github.com/metabase/metabase/blob/d894d8f2cf98911921dde0ca94962113a77e1b34/src/metabase/api/dashboard.clj#L558)):*
> ```Clojure
> (defn- update-cards! [dashboard cards]
>   (check-updated-parameter-mapping-permissions (:id dashboard) cards)
>   ;; transform the card data to the format of the DashboardCard model
> ```

This had me confused for a second, thinking a card was somehow being transformed into a dashboardcard



_Suggestion:_
```Clojure
  ;; transform the dashcard data to the format of the DashboardCard model
```
___
*[`src/metabase/api/dashboard.clj` line 565 at r14](https://reviewable.io/reviews/metabase/metabase/29641#-NSu5gfA0IV9AGWiW4ci:-NSu5gfB9lb9f9SDC3cq:bethdq3) ([raw file](https://github.com/metabase/metabase/blob/d894d8f2cf98911921dde0ca94962113a77e1b34/src/metabase/api/dashboard.clj#L565)):*
> ```Clojure
> 
> (defn- delete-cards! [dashboard dashcard-ids]
>   (when-let [dashboard-cards (t2/select DashboardCard :id [:in dashcard-ids])]
> ```

`dashboard-cards` will always be non-nil. You need to check if it's empty
___
*[`src/metabase/util/malli/schema.clj` line 53 at r14](https://reviewable.io/reviews/metabase/metabase/29641#-NSu9xfi7JwCgzgrgCmQ:-NSu9xfi7JwCgzgrgCmR:undefined) ([raw file](https://github.com/metabase/metabase/blob/d894d8f2cf98911921dde0ca94962113a77e1b34/src/metabase/util/malli/schema.clj#L53)):*
> ```Clojure
> 
> (def NegativeInt
>   "Schema representing an integer than must also be greater than zero."
> ```





_Suggestion:_
```Clojure
  "Schema representing an integer than must be less than zero."
```
___
*[`test/metabase/api/dashboard_test.clj` line 1198 at r14](https://reviewable.io/reviews/metabase/metabase/29641#-NSuA6iB2aWtUOfMfRWf:-NSuA6iB2aWtUOfMfRWg:bw0rory) ([raw file](https://github.com/metabase/metabase/blob/d894d8f2cf98911921dde0ca94962113a77e1b34/test/metabase/api/dashboard_test.clj#L1198)):*
> ```Clojure
>       (let [mappings [{:parameter_id "_CATEGORY_ID_"
>                        :target       [:dimension [:field (mt/id :venues :category_id) nil]]}]]
>         ;; TODO -- check series as well?
> ```

TODO later? Or now?


<!-- Sent from Reviewable.io -->

metamben reviewed (approved):
> Reviewed 26 of 26 files at r13, 2 of 2 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29641)* status: all files reviewed, 15 unresolved discussions (waiting on @alxnddr, @calherries, @EmmadUsmani, and @qnkhuat)

___
*[`src/metabase/api/dashboard.clj` line 521 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160950697:-NSulWhn5LH-XH5gF7t9:be0l5ae) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L521)):*
<details><summary><i>Previously, calherries (Cal Herries) wrote…</i></summary><blockquote>

How about `diff-by-id`, since it's basically doing a `clojure.data.diff` on the ids
</blockquote></details>

I wasn't discussing the name of the function or its output. I was discussing the names `current-change-ids` and `new-change-ids`. Those sound like as if "change" were an entity of our domain, but I don't think they are, and so I have no intuition what they ought to mean.
___
*[`src/metabase/api/dashboard.clj` line 523 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160951380:-NSumAHa4y-TsTyo-yaH:b8w6uoi) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L523)):*
<details><summary><i>Previously, qnkhuat (Ngoc Khuat) wrote…</i></summary><blockquote>

updated d894d8f2cf
</blockquote></details>

I would've used `comp` for `to-create` and `to-update` too.


<!-- Sent from Reviewable.io -->

qnkhuat commented:
*[`src/metabase/api/dashboard.clj` line 541 at r14](https://reviewable.io/reviews/metabase/metabase/29641#-NSu8OSg1nJQqRjdawj9:-NSv7XZPDW2L2ql37E-G:b5xpnfk) ([raw file](https://github.com/metabase/metabase/blob/d894d8f2cf98911921dde0ca94962113a77e1b34/src/metabase/api/dashboard.clj#L541)):*
<details><summary><i>Previously, calherries (Cal Herries) wrote…</i></summary><blockquote>

Also would be nice to use `dashcards` instead of `cards`
</blockquote></details>

updated all create/update/delete functions to have `cards` instead of `dashcards` now.



_Code quote:_
```Clojure
create-cards!
```


<!-- Sent from Reviewable.io -->

qnkhuat commented:
*[`test/metabase/api/dashboard_test.clj` line 1198 at r14](https://reviewable.io/reviews/metabase/metabase/29641#-NSuA6iB2aWtUOfMfRWf:-NSv8E3yByUU2rF0AyHm:b-z2wp9j) ([raw file](https://github.com/metabase/metabase/blob/d894d8f2cf98911921dde0ca94962113a77e1b34/test/metabase/api/dashboard_test.clj#L1198)):*
<details><summary><i>Previously, calherries (Cal Herries) wrote…</i></summary><blockquote>

TODO later? Or now?
</blockquote></details>

later since it's not the scope of this PR. I think it's show up in the changes because I moved the fn.


<!-- Sent from Reviewable.io -->

qnkhuat commented:
*[`src/metabase/api/dashboard.clj` line 523 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160951380:-NSv95dj0TF2RduEJ-RA:by6tqu) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L523)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

I would've used `comp` for `to-create` and `to-update` too.
</blockquote></details>

updated in a67ad0ebeb


<!-- Sent from Reviewable.io -->

qnkhuat commented:
*[`src/metabase/api/dashboard.clj` line 540 at r14](https://reviewable.io/reviews/metabase/metabase/29641#-NSu6iXA0uqgltt-S-YQ:-NSv9FRECvFgb3YXflom:bptt9q0) ([raw file](https://github.com/metabase/metabase/blob/d894d8f2cf98911921dde0ca94962113a77e1b34/src/metabase/api/dashboard.clj#L540)):*
<details><summary><i>Previously, calherries (Cal Herries) wrote…</i></summary><blockquote>

How about renaming this to `create-dashcards`? I think it's passable to use `cards` instead of `dashcards` in a context when it's not ambiguous. But top-level functions lack context by themselves. If you have functions calling functions calling functions where everything is named `card` and not `dashcard`, I'd argue it slows down reading because you have to check whether a thing called `card` is really a `card` or a `dashcard`.
</blockquote></details>

yes! thanks for reminding me. I did a `/s/cards/dashcards` for these create/update/delete methods a67ad0ebeb


<!-- Sent from Reviewable.io -->

qnkhuat commented:
*[`src/metabase/api/dashboard.clj` line 558 at r14](https://reviewable.io/reviews/metabase/metabase/29641#-NSu8J7v43V-WdwlHZk9:-NSv9W_R8LidTAvFwKy8:b-wbwynq) ([raw file](https://github.com/metabase/metabase/blob/d894d8f2cf98911921dde0ca94962113a77e1b34/src/metabase/api/dashboard.clj#L558)):*
<details><summary><i>Previously, calherries (Cal Herries) wrote…</i></summary><blockquote>

This had me confused for a second, thinking a card was somehow being transformed into a dashboardcard
</blockquote></details>

a67ad0ebeb


<!-- Sent from Reviewable.io -->

qnkhuat commented:
*[`src/metabase/api/dashboard.clj` line 565 at r14](https://reviewable.io/reviews/metabase/metabase/29641#-NSu5gfA0IV9AGWiW4ci:-NSv9XixBN86aawPQoFV:b-vk410d) ([raw file](https://github.com/metabase/metabase/blob/d894d8f2cf98911921dde0ca94962113a77e1b34/src/metabase/api/dashboard.clj#L565)):*
<details><summary><i>Previously, calherries (Cal Herries) wrote…</i></summary><blockquote>

`dashboard-cards` will always be non-nil. You need to check if it's empty
</blockquote></details>

yes a67ad0ebeb


<!-- Sent from Reviewable.io -->

qnkhuat reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29641)* status: 48 of 50 files reviewed, 14 unresolved discussions (waiting on @alxnddr, @calherries, @EmmadUsmani, and @metamben)

___
*[`src/metabase/util/malli/schema.clj` line 53 at r14](https://reviewable.io/reviews/metabase/metabase/29641#-NSu9xfi7JwCgzgrgCmQ:-NSv9ZJ9DKkLSIPUW1cj:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/d894d8f2cf98911921dde0ca94962113a77e1b34/src/metabase/util/malli/schema.clj#L53)):*
> ```Clojure
> 
> (def NegativeInt
>   "Schema representing an integer than must also be greater than zero."
> ```

Done.


<!-- Sent from Reviewable.io -->

EmmadUsmani committed:
(hopefully) fix `line_chart.cy.spec.js`

metamben reviewed (changes requested):
> Reviewed 1 of 3 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29641)* status: 48 of 50 files reviewed, 15 unresolved discussions (waiting on @alxnddr, @calherries, @EmmadUsmani, and @qnkhuat)

___
*[`src/metabase/api/dashboard.clj` line 521 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160950697:-NSwMtdg5ieB-Fkkmam_:b2g9gcj) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L521)):*
<details><summary><i>Previously, qnkhuat (Ngoc Khuat) wrote…</i></summary><blockquote>

I see, can you review the new changes? I did a `/s/changes/items` and use the `comp + :id` pattern now.
</blockquote></details>

I'm seeing older versions, have some changes got lost? The changes you describe are great, but I can't see them in GH.


<!-- Sent from Reviewable.io -->

EmmadUsmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29641)* status: 48 of 50 files reviewed, 15 unresolved discussions (waiting on @alxnddr, @calherries, @EmmadUsmani, @metamben, and @qnkhuat)

___
*[`src/metabase/api/dashboard.clj` line 521 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160950697:-NSwOieJ8NjAhXhmYYC8:bokim5) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L521)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

I'm seeing older versions, have some changes got lost? The changes you describe are great, but I can't see them in GH.
</blockquote></details>

Oops I think that was my doing. I force pushed a new commit but I think I forgot to pull the updated branch with Ngoc's new changes before I did it. Ngoc should be able to just push that commit again on top of mine.


<!-- Sent from Reviewable.io -->

qnkhuat committed:
- rename functions with cards -> dashcards
    - update classify-changes to use items as name instead
    - update classify-changes to use remove + comp pattern
    - some docstring cleanup

qnkhuat commented:
@metamben I've re-pushed the lost commits

metamben reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29641)* status: 48 of 50 files reviewed, 14 unresolved discussions (waiting on @alxnddr, @calherries, @EmmadUsmani, and @qnkhuat)

___
*[`src/metabase/api/dashboard.clj` line 523 at r5](https://reviewable.io/reviews/metabase/metabase/29641#gh-1160951380:-NT-8aqA32VAf11jpNhG:bv0o380) ([raw file](https://github.com/metabase/metabase/blob/af798f3f1cde110ee4fcb532b0f81ae7db7ec96b/src/metabase/api/dashboard.clj#L523)):*
<details><summary><i>Previously, qnkhuat (Ngoc Khuat) wrote…</i></summary><blockquote>

updated in a67ad0ebeb
</blockquote></details>

Sorry, I wasn't clear. `filter` is better than `remove` unless you have to use `not`. That is
```
to-create          (filter (comp neg? :id) new-items)
to-update          (filter (comp pos? :id) new-items)
```
are better, IMHO.


<!-- Sent from Reviewable.io -->
